src/thinblock.cpp

```diff
@@ -261,84 +261,90 @@ bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strComm
         return error("%s message received from a non thinblock node, peer=%d", strCommand, pfrom->GetId());
     }
 
+    bool fAlreadyHave = false;
     int nSizeThinBlock = vRecv.size();
+    CInv inv(MSG_BLOCK, uint256());
+
     CXThinBlock thinBlock;
     vRecv >> thinBlock;
 
-    // Message consistency checking
-    if (!IsThinBlockValid(pfrom, thinBlock.vMissingTx, thinBlock.header))
     {
         LOCK(cs_main);
-        Misbehaving(pfrom->GetId(), 100);
-        return error("Invalid %s received", strCommand);
-    }
 
-    // Is there a previous block or header to connect with?
-    {
-        LOCK(cs_main);
-        uint256 prevHash = thinBlock.header.hashPrevBlock;
-        BlockMap::iterator mi = mapBlockIndex.find(prevHash);
-        if (mi == mapBlockIndex.end())
+        // Message consistency checking (FIXME: some redundancy here with AcceptBlockHeader)
+        if (!IsThinBlockValid(pfrom, thinBlock.vMissingTx, thinBlock.header))
         {
-            Misbehaving(pfrom->GetId(), 10);
-            return error("%s from peer %s (%d) will not connect, unknown previous block %s", strCommand,
-                pfrom->addrName.c_str(), pfrom->id, prevHash.ToString());
-        }
-        CBlockIndex *pprev = mi->second;
-        CValidationState state;
-        if (!ContextualCheckBlockHeader(thinBlock.header, state, pprev))
-        {
-            // Thin block does not fit within our blockchain
             Misbehaving(pfrom->GetId(), 100);
-            return error("%s from peer %s (%d) contextual error: %s", strCommand, pfrom->addrName.c_str(), pfrom->id,
-                state.GetRejectReason().c_str());
+            LogPrintf("Received an invalid %s from peer %s (%d)\n",
+                      strCommand, pfrom->addrName.c_str(), pfrom->id);
+            return true;
         }
-    }
 
-    CInv inv(MSG_BLOCK, thinBlock.header.GetHash());
-    bool fAlreadyHave = false;
+        CValidationState state;
+        CBlockIndex *pIndex = NULL;
+        if (!AcceptBlockHeader(thinBlock.header, state, Params(), &pIndex))
+        {
```

Comment:
if we get an out of order thinblock somehow won't we end up invalidating this header and then the chain will  no longer extend?